Sage/carbon
# fix(select): ensure select list always has correct height to fit content

Use a callback ref to ensure that the the code that calculates the list height is ran whenever the
DOM node is attached.

fix #5341

### Proposed behaviour



Select list should always be the correct height to show all its options, with the correct padding/margin, no matter how many times the text wraps.

### Current behaviour

In some circumstances (needs few enough options for there not to be a scrollbar), options with long text that wraps cause the list height to be too small and some parts of some options therefore are not visible.



### Checklist




#### QA


### Additional context



There's a bit of history behind this bug/change that reviewers may find useful.

This was first reported as issue #4489.
Fixed in [this PR](https://github.com/Sage/carbon/pull/4573) by adding `listRef.current` to the `useLayoutEffect`'s dependency array - this works because the height needs to be recalculated when the select list container appears in the DOM, which wasn't happening previously.
This fix was removed in [this PR](https://github.com/Sage/carbon/pull/4949), as part of a much wider set of changes, it seems without any comment at all!

While restoring the dependency fixes the problem, I assume the reason for the removal was that it causes lint warnings saying that a `ref.current` shouldn't be used as an effect dependency - effects only run after a render but an update to a ref does not cause a rerender to happen, so this only works when we are fortunate enough for a prop or state change to happen at about the same time, and I didn't want to rely on this.

It seems the way to guarantee code to run when a ref is updated is to use a callback ref, so that is what I have implemented here.

### Testing instructions



The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.


Codesandbox from original issue: https://codesandbox.io/s/select-options-height-resize-uy2shp?file=/src/index.js

I have changed the "Select/Test" story to a configuration that is more likely to trigger the bug, so this can also be tested (and I have added a new Cypress test to hopefully catch any regressions in this behaviour).
